daemon/access.go

```diff
@@ -0,0 +1,164 @@
+)
+
+func checkPolkitActionImpl(r *http.Request, ucred *ucrednet, action string) accessResult {
+	var flags polkit.CheckFlags
+	allowHeader := r.Header.Get(client.AllowInteractionHeader)
+	if allowHeader != "" {
+		if allow, err := strconv.ParseBool(allowHeader); err != nil {
+			logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+		} else if allow {
+			flags |= polkit.CheckAllowInteraction
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	switch authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err {
+	case nil:
+		if authorized {
+			// polkit says user is authorised
+			return accessOK
+		}
+	case polkit.ErrDismissed:
+		return accessCancelled
+	default:
+		logger.Noticef("polkit error: %s", err)
+	}
+	return accessUnauthorized
+}
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+// openAccess allows requests without authentication, provided they
+// have peer credentials and were not received on snapd-snap.socket
+type openAccess struct{}
+
+func (ac openAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred == nil {
+		return accessForbidden
+	}
+	// Forbid access from snapd-snap.socket
+	if ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	return accessOK
+}
+
+// authenticatedAccess allows requests from authenticated users,
+// provided they were not received on snapd-snap.socket
+//
+// A user is considered authenticated if they provide a macaroon, are
+// the root user according to peer credentials, or granted access by
+// Polkit.
+type authenticatedAccess struct {
+	Polkit string
+}
+
+func (ac authenticatedAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Require peer credentials
+	if ucred == nil {
+		return accessForbidden
+	}
+
+	// Forbid access from snapd-snap.socket
+	if ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	if user != nil {
+		return accessOK
+	}
+
+	if ucred.uid == 0 {
+		return accessOK
+	}
+
+	// We check polkit last because it may result in the user
+	// being prompted for authorisation.  This should be avoided
+	// if access is otherwise granted.
+	if ac.Polkit != "" {
+		return checkPolkitAction(r, ucred, ac.Polkit)
+	}
+
+	return accessUnauthorized
+}
+
+// rootAccess allows requests from the root uid, provided they
+// were not received on snapd-snap.socket
+type rootAccess struct{}
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred == nil {
+		return accessForbidden
```

Comment:
this is not tested.  There's  a bit of pointless variability btw on how the various checkAccess checks the same things. In n general all them except snapAccess do the same check as openAccess, so maybe the latter should also exist as a helper function that is called first by all of them.